shell/browser/node_service_host_impl.cc

```diff
@@ -6,27 +6,28 @@
 
 #include <utility>
 
+#include "content/browser/utility_process_host.h"
+#include "content/public/browser/child_process_data.h"
 #include "shell/browser/api/electron_api_utility_process.h"
 
 namespace electron {
 
-NodeServiceHostImpl::NodeServiceHostImpl() = default;
+NodeServiceHostImpl::NodeServiceHostImpl(
+    content::UtilityProcessHost* utility_process_host)
+    : utility_process_host_(utility_process_host) {}
 
 NodeServiceHostImpl::~NodeServiceHostImpl() = default;
 
-void NodeServiceHostImpl::BindReceiver(
-    mojo::PendingReceiver<node::mojom::NodeServiceHost> receiver) {
-  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
-  receivers_.Add(this, std::move(receiver));
-}
-
 void NodeServiceHostImpl::ReceivePostMessage(
-    base::ProcessId pid,
     blink::TransferableMessage message) {
   DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
-  auto utility_process_wrapper = api::UtilityProcessWrapper::FromProcessId(pid);
-  if (utility_process_wrapper)
-    utility_process_wrapper->ReceivePostMessage(std::move(message));
+  if (utility_process_host_) {
```

Comment:
Do we need to observe the UtilityProcessHost to catch its destruction and null out our pointer here?

Reply:
Host only gets torn down after the child process terminates, when the child process terminates it will close the remote and that would cause this class to be destroyed since it is managed by the receiver lifetime. So we should never encounter a case were this class outlives `UtilityProcessHost`